¡Hola!

Mi nombre es Tonatiuh Cruz. Me complace revisar tu proyecto hoy.

Al identificar cualquier error inicialmente, simplemente los destacaré. Te animo a localizar y abordar los problemas de forma independiente como parte de tu preparación para un rol como data-scientist. En un entorno profesional, tu líder de equipo seguiría un enfoque similar. Si encuentras la tarea desafiante, proporcionaré una pista más específica en la próxima iteración.

Encontrarás mis comentarios a continuación - **por favor no los muevas, modifiques o elimines**.

Puedes encontrar mis comentarios en cajas verdes, amarillas o rojas como esta:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Éxito. Todo está hecho correctamente.
</div>

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Observaciones. Algunas recomendaciones.
</div>

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Necesita corrección. El bloque requiere algunas correcciones. El trabajo no puede ser aceptado con comentarios en rojo.
</div>

Puedes responderme utilizando esto:

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante.</b> <a class="tocSkip"></a>
</div>

# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
    * [Objetivo](#Objetivo)
    * [Etapas](#Etapas)

* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
    * [Problemas detectados](#Problemas-detectados)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
        * [2.4.1 Mis Conclusiones](#Mis_data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypothesis)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
* [Conclusiones](#end)

<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

La tabla de contenidos está bien estructurada, pero sería útil si estuviera enlazada a las secciones correspondientes, de manera que al hacer clic se pueda acceder directamente a cada una. Esto facilitaría la navegación.

Un consejo útil: si etiquetas correctamente todas las secciones con sus respectivos números (#), puedes generar automáticamente la tabla de contenidos desde Jupyter Lab. Simplemente ve a la pestaña de herramientas y haz clic en el botón de los puntos y barras (Table of contents). Esto te generará una tabla de contenidos enlazada y visualmente atractiva. El botón está ubicado a la derecha del botón "Validate".
</div>



## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar la hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo 
Prueba la hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.

### Etapas 
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar la hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos.
 2. Preprocesamiento de datos.
 3. Prueba de hipótesis.








[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [1]:
# Importar pandas. con esta libreria vamos a poder realizar el sprint 
import pandas as pd

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [27]:
# Leer el archivo y almacenarlo en df
df = pd.read_csv('/datasets/music_project_en.csv')


Muestra las 10 primeras filas de la tabla:

In [3]:
# Obtener las 10 primeras filas de la tabla df
# es importante leer los archivos, al ver las informaciones se puede verificar si los nombres se parecen a los datos y es es muy impotante 
df.head(10)

userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

Obtén la información general sobre la tabla con un comando. Conoces el método que muestra la información general que necesitamos.

In [4]:
# Obtener la información general sobre nuestros datos
# viendo lasinformaciones generales del  DataFrame con info podemos ver 
df.info()



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Almacenan los mismos tipos de datos: `object`.

Según la documentación:
- `' userID'`: identificador del usuario o la usuaria;-columna userID: 65,079 entradas no nulas, tipo de dato object.
- `'Track'`: título de la canción;-columna Track: 63,736 entradas no nulas, tipo de dato object. lineas a eliminar 1,343 
- `'artist'`: nombre del artista;-columna artist: 57,512 entradas no nulas, tipo de dato object. lineas a eliminar 7,567
- `'genre'`: género de la pista;-columna genre: 63,881 entradas no nulas, tipo de dato object. lineas a eliminar 1,198
- `'City'`: ciudad del usuario o la usuaria;-City: 65,079 entradas no nulas, tipo de dato object.
- `'time'`: la hora exacta en la que se reprodujo la canción;-columna time: 65,079 entradas no nulas, tipo de dato object.
- `'Day'`: día de la semana.-columna Day: 65,079 entradas no nulas, tipo de dato object.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas, otros en minúsculas.
2. Hay espacios en algunos encabezados.
3. `Detecta el tercer problema por tu cuenta y descríbelo aquí`.
### Problemas detectados
3. convertir todos los nombres de las columnas a minúsculas para la consistencia. fuimos capacitados en corregir estos casos ya que, es coun tener errores al llamar estas columnas y por convencion estos nombre no se usan con los  
4. Ademas si el nombre tiene varias palabras, utiliza snake_case como norma de python

### Escribe observaciones de tu parte. Estas son algunas de las preguntas que pueden ser útiles: <a id='data_review_conclusions'></a>



#### 2.4.1 Mis Conclusiones](#Mis_data_preprocessing_conclusions)
`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`

-Las filas almacenan información sobre eventos individuales de escucha de música por usuarios específicos. indicando el usuario, el titulo de la cancion, el artista, el genero musical al que pertenece, la ciudad donde se reproducjo la cancion, el tiempo que duro la reproduccion y el dia de la reproduccion.

-user_id: es un identificador único para cada usuario, posiblemente en formato alfanumérico. esta columna fua cambiada nace de userid.

-track: el título de la canción que fue reproducida. este es un campo especial pues pueden existir canciones del mismo nombre de artistas diferentes.

-artist: el nombre del artista que interpretó la canción. cada artista busca tener un nombre unico en este camopo no tendremos situaciones como las del campo track.

-genre: el género musical de la pista reproducida. es importante la clasificacion de los geners esto ayudara a conocer los genenros mas populares por cada ciudad.

-city: la ciudad donde el usuario escuchó la canción. este campo nos va a indicar la comparativa entre las dos ciudades que estudiaremos.

-time: la hora exacta cuando la canción fue reproducida, probablemente en formato de hora.

-day: el día de la semana cuando la canción fue reproducida. con este campo tambien realizaremos comparaciones para determinar que tan importante puede ser.


`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestra hipótesis o necesitamos más información?`

-si, hay una cantidad significativa de datos (65,079 filas de eventos de escucha) que deberían permitir analizar las diferencias en la actividad de escucha entre las dos ciudades y en diferentes días de la semana. aunque existen lineas duplicadas, campos vacios.

-es importante señala rque, la suficiencia de los datos no solo depende de la cantidad sino también de la calidad. es para el analisis que realizaremos esencial que los datos sean representativos de las poblaciones de interes. 

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`

-valores ausentes: Hemos observado que hay valores nulos en las columnas track, artist, y genre. Esto podría afectar el análisis relacionado con las preferencias de música si una proporción significativa de esos datos está ausente.

-duplicados: aun no sea verificado si hay filas duplicadas, lo cual sería necesario hacer para garantizar la precisión del análisis.

-tipos de datos incorrectos: todas las columnas se han leído como tipo de dato object. Es posible que necesitemos convertir la columna time a un tipo de dato de tiempo si queremos realizar algún análisis que dependa del tiempo exacto de los eventos. esto se realizo al correr ¨df.info()¨

-viendo los datos y recordando los ejercicios podemos deducir que hay que verificar la consistencia en los datos; como la uniformidad en los nombres de artistas, pistas y géneros, para asegurarnos de que no haya errores tipográficos o variantes en los nombres que deberían ser unificados.


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo! Excelentes observaciones de tu parte. Solamente que en la primer pregunta debes de comentar un poco sobre los tipo de variables que tenemos en nuestra base de datos.</div>

[Volver a Contenidos](#back)

## 3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla (los nombres de las columnas):

In [5]:
# Muestra los nombres de las columnas
df.columns.tolist()

['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day']

Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* Todos los caracteres deben ser minúsculas.
* Elimina los espacios.
* Si el nombre tiene varias palabras, utiliza snake_case.

Anteriormente, aprendiste acerca de la forma automática de cambiar el nombre de las columnas. Vamos a aplicarla ahora. Utiliza el bucle for para iterar sobre los nombres de las columnas y poner todos los caracteres en minúsculas. Cuando hayas terminado, vuelve a mostrar los encabezados de la tabla:

In [6]:
# Bucle en los encabezados poniendo todo en minúsculas
df.columns = [col.lower() for col in df.columns]
# para ver nuevamente los encabezados en minusculas df.columns.tolist()

Ahora, utilizando el mismo método, elimina los espacios al principio y al final de los nombres de las columnas e imprime los nombres de las columnas nuevamente:

In [7]:
# Bucle en los encabezados eliminando los espacios
df.columns = [col.strip() for col in df.columns]
# para ver nuevamente los encabezados sin espacios df.columns.tolist()

Necesitamos aplicar la regla de snake_case a la columna `userid`. Debe ser `user_id`. Cambia el nombre de esta columna y muestra los nombres de todas las columnas cuando hayas terminado.

In [8]:
# Cambiar el nombre de la columna "userid"
df.rename(columns={'userid': 'user_id'}, inplace=True)
# para ver nuevamente el cambio en la columna user_id df.columns.tolist()

Comprueba el resultado. Muestra los encabezados una vez más:

In [9]:
# Comprobar el resultado: la lista de encabezados
df.columns.tolist()

['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day']

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo! Ajustaste todos los elementos</div>

[Volver a Contenidos](#back)

### Valores ausentes <a id='missing_values'></a>
 Primero, encuentra el número de valores ausentes en la tabla. Debes utilizar dos métodos en una secuencia para obtener el número de valores ausentes.

In [10]:
# Calcular el número de valores ausentes
# utilizando isnull en el dataframe
missing_values = df.isnull().sum()
#imprimir los valores 
print(missing_values)

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en las columnas `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Como mostramos anteriormente en las lecciones, la mejor forma de hacerlo es crear una lista que almacene los nombres de las columnas donde se necesita el reemplazo. Luego, utiliza esta lista e itera sobre las columnas donde se necesita el reemplazo haciendo el propio reemplazo.

In [11]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

columns_to_replace = ['track', 'artist', 'genre']
for column in columns_to_replace:
    df[column].fillna('unknown', inplace=True)


Ahora comprueba el resultado para asegurarte de que después del reemplazo no haya valores ausentes en el conjunto de datos. Para hacer esto, cuenta los valores ausentes nuevamente.

In [12]:
# Contar valores ausentes

df.isnull().sum()
# verificando los cambios recontando los valores ausentes para asegurarnos de que han sido reemplazados correctamente. vemoso que todos estan en 0

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo! Ajustaste todos los elementos</div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla. Una vez más, debes aplicar dos métodos en una secuencia para obtener la cantidad de duplicados explícitos.

In [26]:
# Contar duplicados explícitos
duplicate_count = df.duplicated().sum()
duplicate_count
# como nos fue enseñado primero hay que usar duplicated(debuelve true) para identificar las filas duplicadas y luego sum(suamrisa los true) para contarlas

0

Ahora, elimina todos los duplicados. Para ello, llama al método que hace exactamente esto.

In [14]:
# Eliminar duplicados explícitos
# llamando drop_duplicates() en el dataframe podemos eliminar las finalas que tiene duplicado ¨explícitos¨
df = df.drop_duplicates()


Comprobemos ahora si eliminamos con éxito todos los duplicados. Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [15]:
# Comprobar de nuevo si hay duplicados

df.info()

# con info(), verificamos la cantidad de filas restantes para confirmar la eliminación de duplicados

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61253 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   user_id  61253 non-null  object
 1   track    61253 non-null  object
 2   artist   61253 non-null  object
 3   genre    61253 non-null  object
 4   city     61253 non-null  object
 5   time     61253 non-null  object
 6   day      61253 non-null  object
dtypes: object(7)
memory usage: 3.7+ MB


Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para ello:
* Extrae la columna `genre` del DataFrame.
* Llama al método que devolverá todos los valores únicos en la columna extraída.


In [16]:
# Inspeccionar los nombres de géneros únicos

# buscar en la columna 'genre' con unique() para obtener los valores unicos ordenados alfabeticamente esto ayuda a encontar valores mal escrito y encontrar erroes de escritura
unique_genres = sorted(df['genre'].unique())
unique_genres



['acid',
 'acoustic',
 'action',
 'adult',
 'africa',
 'afrikaans',
 'alternative',
 'ambient',
 'americana',
 'animated',
 'anime',
 'arabesk',
 'arabic',
 'arena',
 'argentinetango',
 'art',
 'audiobook',
 'avantgarde',
 'axé',
 'baile',
 'balkan',
 'beats',
 'bigroom',
 'black',
 'bluegrass',
 'blues',
 'bollywood',
 'bossa',
 'brazilian',
 'breakbeat',
 'breaks',
 'broadway',
 'cantautori',
 'cantopop',
 'canzone',
 'caribbean',
 'caucasian',
 'celtic',
 'chamber',
 'children',
 'chill',
 'chinese',
 'choral',
 'christian',
 'christmas',
 'classical',
 'classicmetal',
 'club',
 'colombian',
 'comedy',
 'conjazz',
 'contemporary',
 'country',
 'cuban',
 'dance',
 'dancehall',
 'dancepop',
 'dark',
 'death',
 'deep',
 'deutschrock',
 'deutschspr',
 'dirty',
 'disco',
 'dnb',
 'documentary',
 'downbeat',
 'downtempo',
 'drum',
 'dub',
 'dubstep',
 'eastern',
 'easy',
 'electronic',
 'electropop',
 'emo',
 'entehno',
 'epicmetal',
 'estrada',
 'ethnic',
 'eurofolk',
 'european',
 'expe

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, crea una función llamada `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=`: esta es una lista que contiene todos los valores que necesitas reemplazar.
* `correct_genre=`: este es un string que vas a utilizar como reemplazo.

Como resultado, la función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplazar cada valor de la lista `wrong_genres` por el valor en `correct_genre`.

Dentro del cuerpo de la función, utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos, extrae la columna `'genre'` y aplica el método `replace` para hacer correcciones.

In [17]:
# Función para reemplazar duplicados implícitos
# definimos la funcion para reemplazar genero musical incorrecto con un ciclo for
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

wrong_genres_list = ['hip', 'hop', 'hip-hop']
correct_genre_name = 'hiphop'

#reemplazamos los valores
replace_wrong_genres(wrong_genres_list, correct_genre_name)

# Para verificar los cambios
sorted_unique_genres = sorted(df['genre'].unique())
print(sorted_unique_genres)

['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

Ahora, llama a `replace_wrong_genres()` y pásale tales argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [18]:
# Eliminar duplicados implícitos

# crear de la función replace_wrong_genres para reemplazar generos incorrectos
def replace_wrong_genres(wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

# cear lista de los generos incorrectos y en la funcion correct_genre_name pasar el genero correcto
wrong_genres_list = ['hip', 'hop', 'hip-hop']
correct_genre_name = 'hiphop'

# Llamar para que se ejecute la funcion realizando las correcciones
replace_wrong_genres(wrong_genres_list, correct_genre_name)



Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [19]:
# Comprobación de duplicados implícitos

# Extraer y mostrar los géneros únicos para verificar las correcciones
sorted_unique_genres_after_correction = sorted(df['genre'].unique())
print(sorted_unique_genres_after_correction)


['acid', 'acoustic', 'action', 'adult', 'africa', 'afrikaans', 'alternative', 'ambient', 'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena', 'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile', 'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues', 'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks', 'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean', 'caucasian', 'celtic', 'chamber', 'children', 'chill', 'chinese', 'choral', 'christian', 'christmas', 'classical', 'classicmetal', 'club', 'colombian', 'comedy', 'conjazz', 'contemporary', 'country', 'cuban', 'dance', 'dancehall', 'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr', 'dirty', 'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum', 'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop', 'emo', 'entehno', 'epicmetal', 'estrada', 'ethnic', 'eurofolk', 'european', 'experimental', 'extrememetal', 'fado', 'film', 'fitness', 'flamenco', 'folk', 'folklor

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo! Remplazaste los generos incorrectos y eliminaste los duplicados.</div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

`Describe brevemente lo que has notado al analizar duplicados, cómo abordaste sus eliminaciones y qué resultados obtuviste.`


### Resultado de las Observaciones 

Hemos podido realizar ejercicios en los que pudimos observar duplicados de dos tipos: explícitos e implícitos, los cuales podemos describir de la siguiente manera:

#### duplicados explicitos:
- observacion: detecte duplicados explicitos utilizando el metodo `duplicated().sum()` que revelo la presencia de filas completamente identicas.
- limpiesa: se realizo una eliminacion de los duplicados explicitos utilizando el metodo `drop_duplicates()` para eliminar estas filas duplicadas del DataFrame. 
- resultante: la eliminacion las lineas duplicadas con esta condicion, lo cual es esencial para un análisis estadístico válido.

#### duplicados implicitos:
- observación: en la columna `genre`, existian duplicados implicitos, como 'hip', 'hop', y 'hip-hop' que en realidad se refieren al mismo genero musical.
- limpiesa: se realizo una eliminacion de los duplicado utilizando `replace_wrong_genres()` que recibe una lista de generos incorrectos y  genero correcto para reemplazar los nombres incoherentes o mal escritos. Esta función itera sobre la columna `genre` y reemplaza los términos incorrectos con un termino unificado, en este caso, 'hiphop'.
- resultante: tras aplicar esta funcion, se eliminaron los nombres duplicados para reflejar de manera precisa y coherente las categorias musicales.

 conclusion:
La limpieza de duplicados tanto explicitos como implicitos esta dando coherencia y mejoras significativas en las informaciones y en la calidad del dataset. 
Estamos practicando lo ejerciciones del sprint y preparando los datos para un análisis de datos más confiable. 



[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypothesis'></a>

### Hipótesis: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

El primer paso es evaluar la actividad del usuario en cada ciudad. Recuerda las etapas dividir-aplicar-combinar de las que hablamos anteriormente en la lección. Tu objetivo ahora es agrupar los datos por ciudad, aplicar el método apropiado para contar durante la etapa de aplicación y luego encontrar la cantidad de canciones reproducidas en cada grupo especificando la columna para obtener el recuento.

A continuación se muestra un ejemplo de cómo debería verse el resultado final:
`df.groupby(by='....')['column'].method()`Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [20]:
# Contar las canciones reproducidas en cada ciudad

# filtramos por los dias especificados
filtered_df = df[df['day'].isin(['Monday', 'Wednesday', 'Friday'])]


# Agrupar por 'city' y 'day' y contar las reproducciones
song_counts_by_city_day = filtered_df.groupby(['city', 'day'])['track'].count().unstack()

song_counts_by_city_day

day          Friday  Monday  Wednesday
city                                  
Shelbyville    5895    5614       7003
Springfield   15945   15740      11056

`Comenta tus observaciones aquí`
*A- en las 2 ciudades, hay variaciones en la cantidad de canciones que se reproducen en cada dia de la semana. Lo que indica que hay un comportamiento diferente de los usuarios en diferentes dias.

B-actividad de cada ciudad:
Springfield: tiene una mayor cantidad de reproducciones en los tres días observados.
Shelbyville.  tiene una menor cantidad de reproducciones en los tres días observados.
tengo las sopechas que la ciudad Springfield tine mas ciudadanos o es muy conocidad.


tendencias durante la semana:
Shelbyville: el Wednesday es cuando mas se reproduce musica, seguido por el Friday y luego el Monday.
En Springfield, el Friday es el día más activo, seguido por el Monday y luego el Wednesday.


Implicaciones para la Hipótesis:
Los datos sugieren que existe una diferencia en la actividad de escucha entre las dos ciudades, apoyando la hipótesis inicial de que podría haber diferencias en el comportamiento de los usuarios en función de la ciudad.
También parece haber un patrón semanal en la actividad de los usuarios, que puede estar influenciado por factores como la rutina de trabajo/semana laboral y actividades de ocio.

Próximos Pasos:
Se podrían realizar análisis adicionales para comprender las razones detrás de las tendencias observadas, como investigar el género de las canciones reproducidas, la demografía de los usuarios o eventos específicos en esos días.
Sería útil comparar estos datos con otros días de la semana para ver si las tendencias observadas son consistentes.

Ahora agrupemos los datos por día de la semana y encontremos el número de canciones reproducidas el lunes, miércoles y viernes. Utiliza el mismo método que antes, pero ahora necesitamos una agrupación diferente.


In [21]:
# Calcular las canciones reproducidas en cada uno de los tres días

# agrupar los datos por 'day' y contar las reproducciones de canciones para cada día
song_counts_by_day = df.groupby('day')['track'].count()

# Mostrar el número de canciones reproducidas para cada uno de los días especificados
print(song_counts_by_day)

day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64


`Comenta tus observaciones aquí`

este teporte nos demuestra:
- Friday es el dia que mas personas se conectan a la aplicacio por tner 21840.

- Monday es el dia que esta en segundo lugar por usuario conectados a la aplicacion 21354.

- Wednesday es el dia que menos personas se coenctan

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la función `number_tracks()` para calcular el número de canciones reproducidas en un determinado día **y** ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'` (lunes).
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [22]:
# Declara la función number_tracks() con dos parámetros: day= y city=.
def number_tracks(day, city):
    # Filtra los datos por día
    day_data = df[df['day'] == day]
    # Almacena las filas del DataFrame donde el valor en la columna 'day' es igual al parámetro day=

    # Filtra las filas donde el valor en la columna 'city' es igual al parámetro city=
    city_data = day_data[day_data['city'] == city]
    # Extrae la columna 'user_id' de la tabla filtrada y aplica el método count()
    track_count = city_data['user_id'].count()
    # Devolve el número de valores de la columna 'user_id'
    return track_count

    
# Ejemplo de uso de la función para contar las canciones reproducidas en Springfield el lunes
# Esto se debe ejecutar en tu entorno local, ya que no se pueden ejecutar ejemplos interactivos aquí
print(number_tracks('Monday', 'Springfield'))

15740


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [23]:
# El número de canciones reproducidas en Springfield el lunes
tracks_springfield_monday = number_tracks('Monday', 'Springfield')
tracks_springfield_monday

15740

In [24]:
# El número de canciones reproducidas en Shelbyville el lunes
tracks_shelbyville_monday = number_tracks('Monday', 'Shelbyville')
tracks_shelbyville_monday

5614

In [25]:
# El número de canciones reproducidas en Springfield el miércoles
tracks_springfield_wednesday = number_tracks('Wednesday', 'Springfield')
tracks_shelbyville_wednesday

NameError: name 'tracks_shelbyville_wednesday' is not defined

In [ ]:
# El número de canciones reproducidas en Shelbyville el miércoles
tracks_shelbyville_wednesday = number_tracks('Wednesday', 'Shelbyville')
tracks_shelbyville_wednesday

In [ ]:
# El número de canciones reproducidas en Springfield el viernes
tracks_springfield_friday = number_tracks('Friday', 'Springfield')
tracks_shelbyville_friday

In [ ]:
# El número de canciones reproducidas en Shelbyville el viernes
tracks_shelbyville_friday = number_tracks('Friday', 'Shelbyville')
tracks_shelbyville_friday

**Conclusiones**

`Comenta si la hipótesis es correcta o se debe rechazar. Explica tu razonamiento.`

#### 1. Observaciones Generales
- la informacion analizada refleja datos individuales de eventos de escucha de música, que inicialmente presentaban problemas de valores ausentes, duplicados y errores tipográficos. Se realizó una limpieza efectuada ha mejorado notablemente la calidad de los datos, aumentando la fiabilidad de los análisis subsiguientes. 

#### 2. comparación del comportamiento del usuario en Springfield y Shelbyville
- Springfield muestra una mayor actividad en términos de reproducciones musicales, lo cual puede sugerir una mayor base de usuarios o un compromiso más fuerte con la plataforma.
- Shelbyville muestra menor actividad, con picos los miércoles. Estas diferencias respaldan la hipótesis inicial de variaciones en la actividad de escucha según la ciudad y el día de la semana.

#### 3. tendencias y patrones semanales
- los datos revelan que el viernes es el día de mayor actividad de escucha, seguido del lunes y luego el miércoles, indicando patrones que podrían estar influenciados por la rutina semanal de trabajo y ocio.

#### 4. analisis detallado utilizando la funcion number_tracks()
- La función number_tracks() ha permitido un análisis detallado del comportamiento del usuario, destacando diferencias notables en la actividad de escucha entre Springfield y Shelbyville durante los días seleccionados.

#### 5. evaluación final de la hipótesis
- La hipótesis propuesta se acepta. 
- Los análisis realizados demuestran diferencias estadísticamente significativas en la actividad de escucha de música entre los usuarios de ambas ciudades, con Springfield liderando en número de reproducciones en general y en los días estudiados, mientras que Shelbyville muestra una tendencia de menor actividad de escucha.

#### 6. explicación y razonamiento
- La comparativa fue realizada en base en un análisis descriptivo detallado y el uso de funciones planteadas en este Sprint las cuales estan diseñadas para evaluar este ejercicio y con estos encontrar las diferencias entre ciudades.
- Se han identificado diferencias notables que corroboran la hipótesis de variabilidad en el comportamiento de escucha basado en la localización urbana.

#### 7. limitaciones y recomendaciones
- El análisis se limitó a datos de tres días específicos de la semana.
- No se consideraron variables demográficas ni la diversidad de géneros musicales escuchados.


#### 8. conclusión reforzada
- Los patrones observados son coherentes y respaldan la afirmación de que el comportamiento de escucha de los usuarios de música online varía significativamente entre Springfield y Shelbyville, tanto a lo largo de la semana como entre las dos ciudades. 




<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>

Muy buen trabajo! Desarrollaste de forma correcta el análisis comparativo. Además creaste la función de manera correcta. </div>

[Volver a Contenidos](#back)

# 4. Conclusiones <a id='end'></a>

## 4.1 `Resume aquí tus conclusiones sobre la hipótesis.`

**Basado en los análisis realizados y las observaciones detalladas, las conclusiones respecto a la hipótesis de que la actividad de escucha musical varía entre los usuarios de Springfield y Shelbyville, y que esta variación es dependiente del día de la semana, son las siguientes:
### 4.1.1 Aceptación de la Hipótesis
- Los datos soportan la hipótesis. Existe una diferencia clara en el comportamiento de escucha musical entre las ciudades estudiadas. Springfield muestra consistentemente más reproducciones musicales que Shelbyville en los días analizados (lunes, miércoles y viernes), lo cual puede reflejar una mayor base de usuarios o un mayor engagement con la plataforma de música.

### 4.1.2 Tendencias Diarias
- Se observaron patrones de consumo específicos por día de la semana, con el viernes siendo el día de mayor actividad de escucha en Springfield, mientras que en Shelbyville es el miércoles. Estos patrones podrían estar relacionados con las rutinas de trabajo y ocio.

### 4.1.3 Relevancia de los Datos
- La cantidad de datos y el enfoque de limpieza de datos han permitido un análisis suficiente para probar la hipótesis, aunque se reconoce que un análisis más extenso en términos de días y la inclusión de variables demográficas y de género musical podría proporcionar una comprensión más rica y matizada.

### 4.1.4 Limitaciones 
- El estudio se centró en tres días específicos(la base de datos solo tiene etos días) y no incluyó un análisis más amplio que pudiera confirmar las tendencias en un período más largo. 
- No se tomaron en cuenta factores como demografía de usuarios o categorías de género musical específicas, hora de reproducción, los cuales podrían influir significativamente en las tendencias de escucha.

### 4.1.5 Recomendaciones para Investigaciones Futuras
- Se sugiere expandir el análisis a todos los días de la semana y considerar variables adicionales para una comprensión más completa del comportamiento del usuario. 
- Esto podría incluir el análisis de géneros musicales preferidos y la relación de estos con eventos culturales o sociales que podrían influir en los patrones de escucha.
- Se sugiere realizar análisis adicionales, incluyendo datos de toda la semana, considerando la demografía de los usuarios y el género musical, para comprender de manera integral las razones detrás de las tendencias observadas.
- Esta conclusión, aunque basada en datos reales, reconoce las limitaciones del alcance del análisis y fomenta una investigación más profunda para una comprensión más holística y precisa de los hábitos de escucha musical.

**Los resultados del análisis indican que la hipótesis es correcta: hay diferencias significativas en la actividad de escucha musical entre los usuarios de Springfield y Shelbyville, variando según el día de la semana. Sin embargo, para comprender completamente el comportamiento de los usuarios y las usuarias, se necesitarían análisis adicionales y más detallados.

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

<div class="alert alert-block alert-warning">
<b>Comentario general</b> <a class="tocSkip"></a>

Rubintel,  hiciste un muy buen trabajo con el proyecto, pudiste resolver casi todos los ejericios de una forma adecuada y se nota que hiciste uso de todas las herramientas aprendidas. Además, considero que te ayudo para aprender el uso de algunas funciones que te sirvan en los siguientes cursos.Solamente verifica lo de la tabla de contenidos.

Exitos en lo que viene, saludos.</div>

[Volver a Contenidos](#back)